### PR TITLE
[CSL-1526] Support CBOR Data Item for TxIn

### PR DESCRIPTION
**Pull request by adinapoli-iohk**
Same as https://github.com/input-output-hk/cardano-sl/pull/1394 , but opened against `master` so that GH don't think I'm changing a crapload of stuff.

### CODE REVIEW

**Comment by domenkozar** [comment]
This one eats 16G of memory while testing and crashes: https://hydra.iohk.io/log/92sfxb7wwgcz2y4qw5jgiwsab8s8hxay-cardano-sl-0.5.1.drv

**Comment by dcoutts** [comment]
@adinapoli-iohk don't forget that we want to use these same utils for all the other "unknown" cases everywhere, and do something similarly nice for the crc case in address. So that U vs U24 distinction should disappear.

**Comment by adinapoli-iohk** [comment]
> I suggest you squash your patches together and rebase so that it merges cleanly into the other branch.

We can always "Squash & Merge" this one, without the need me to rewrite the history (aka washing my hands, leaving GH the burden 😛  )

> So that U vs U24 distinction should disappear.

Aye, this is what already happens [here](https://github.com/input-output-hk/cardano-sl/pull/1404) (Which I need to upgrade with the new combinator names)

File: core/Pos/Binary/Class/Primitive.hs [hunk 1]
```diff
@@ -199,10 +207,20 @@ decodeCborDataItemTag = do
...
+-- | Like `decodeKnownCborDataItem`, but assumes nothing about the Haskell
+-- type we want to deserialise back, therefore it yields the `ByteString`
+-- Tag 24 sorrounded (stripping such tag away).
+-- In CBOR notation, if the data was serialised as:
+-- >>> 24(h'DEADBEEF')
+-- then `decodeUnknownCborDataItem` yields the inner 'DEADBEEF', unchanged.
+decodeUnknownCborDataItem :: D.Decoder s ByteString
+decodeUnknownCborDataItem = do
+    decodeCborDataItemTag
+    decode @ByteString
```
[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply core/Pos/Binary/Class/Primitive.hs]: Oh, missed the existence of `decodeBytes` thanks!

File: core/Pos/Binary/Class/Primitive.hs [hunk 2]
```diff
@@ -181,11 +182,18 @@ biSize = fromIntegral . BS.length . serialize'
...
-encodeCborDataItem bs = E.encodeTag 24 <> encode bs
+encodeKnownCborDataItem :: Bi a => a -> E.Encoding
+encodeKnownCborDataItem = encodeUnknownCborDataItem . serialize'
+
+-- | Like `encodeKnownCborDataItem`, but assumes nothing about the shape of
+-- input object, so that it must be passed as a binary `ByteString` blob.
+-- It's the caller responsibility to ensure the input `ByteString` correspond
+-- indeed to valid, previously-serialised CBOR data.
+encodeUnknownCborDataItem :: ByteString -> E.Encoding
+encodeUnknownCborDataItem = mappend (E.encodeTag 24) . encode
```
[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply core/Pos/Binary/Class/Primitive.hs]: ok 😝

**Comment by adinapoli-iohk** [comment]
The Darwin build timed out, sigh. 😭

**Comment by dcoutts** [comment]
@adinapoli-iohk I merged PR #1404 so I guess this needs a bit of updating now.

**Comment by adinapoli-iohk** [comment]
Thanks @dcoutts , on it today.

**Pull request merged**